### PR TITLE
CASSANDRA-19772: Deprecate option SIDECAR_INSTANCES and replace with SIDECAR_CONTACT_POINTS

### PR DESCRIPTION
**Pull request by yifan-c**
This patch introduces a new option SIDECAR_CONTACT_POINTS for both bulk writer and reader. The option name better describes the purpose, which is to specify the initial contact points to discover the cluster topology. The existing option SIDECAR_INSTANCES are used for the same purpose and it is now deprecated. In addition, it allows including the port value in the addresses when defining SIDECAR_CONTACT_POINTS

Patch by Yifan Cai; Reviewed by Francisco Guerrero for CASSANDRA-19772

### CODE REVIEW

File: cassandra-analytics-core/src/test/spark3/org/apache/cassandra/spark/common/SidecarInstanceFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+        assertSidecarInstance(SidecarInstanceFactory.createFromString("localhost", 9999),
+                              "localhost", 9999);
+        assertSidecarInstance(SidecarInstanceFactory.createFromString("[2024:a::1]", 9999),
+                              "[2024:a::1]", 9999);
+        assertSidecarInstance(SidecarInstanceFactory.createFromString("localhost:8888", 9999),
+                              "localhost", 8888);
+        assertSidecarInstance(SidecarInstanceFactory.createFromString("127.0.0.1:8888", 9999),
+                              "127.0.0.1", 8888);
+        assertSidecarInstance(SidecarInstanceFactory.createFromString("[2024:a::1]:8888", 9999),
```
Comment (frankgh) [inline cassandra-analytics-core/src/test/spark3/org/apache/cassandra/spark/common/SidecarInstanceFactoryTest.java:46]: I think we'll need to remove brackets from ipv6 addresses.

File: cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/utils/MapUtils.java [hunk 2]
```diff
@@ -239,4 +245,35 @@ public static boolean containsKey(Map<String, String> options, String key)
...
+        T deprecatedOptionValue = null;
+        if (options.containsKey(deprecated))
+        {
+            LOGGER.warn("The option: {} is deprecated. Please use {} instead. See option description for details.", deprecated, option);
+            deprecatedOptionValue = resolver.apply(deprecated);
+        }
+
+        if (options.containsKey(option))
+        {
+            LOGGER.info("The option: {} is defined. Favor the value over {}", option, deprecated);
```
Comment (frankgh) [inline cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/utils/MapUtils.java]: info level here? maybe condition it to `deprecatedOptionValue` not `null`?

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkSparkConf.java [hunk 3]
```diff
@@ -146,17 +148,16 @@ public class BulkSparkConf implements Serializable
     protected boolean useOpenSsl;
     protected int ringRetryCount;
     // create sidecarInstances from sidecarInstancesValue and effectiveSidecarPort
-    private final String sidecarInstancesValue;
+    private final String sidecarContactPointsCsv;
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkSparkConf.java]: Not a huge fan of this variable name, can we use `sidecarContactPointsValue` instead?

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkSparkConf.java [hunk 4]
```diff
@@ -240,38 +241,47 @@ protected Set<String> buildBlockedInstances(Map<String, String> options)
...
         return Arrays.stream(split)
-                     .map(hostname -> new SidecarInstanceImpl(hostname, effectiveSidecarPort))
+                     .filter(StringUtils::isNotEmpty)
+                     .map(hostname -> SidecarInstanceFactory.createFromString(hostname, effectiveSidecarPort))
                      .collect(Collectors.toSet());
     }
 
+    // extract port from input if present and use the extract port to create SidecarInstanceImpl
+    // otherwise, the input does not contain port info, and it uses the default port
+    static SidecarInstanceImpl createSidecarInstance(String input, int defaultPort)
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkSparkConf.java]: unused method?

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/SidecarInstanceFactory.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * Create SidecarInstance object by parsing the input string, which is IP address or hostname and optionally includes port
+     * @param input hostname string that can optionally includes the port. If port is present, the defaultPort param is ignored.
+     * @param defaultPort port value used when the input string contains no port
+     * @return SidecarInstanceImpl
+     */
+    public static SidecarInstanceImpl createFromString(String input, int defaultPort)
+    {
+        Preconditions.checkArgument(StringUtils.isNotEmpty(input), "Unable to create sidecar instance from empty input");
```
Comment (frankgh) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/SidecarInstanceFactory.java:47]: I think HostAndPort from guava is an alternative option to use. It is included as a transitive dependency from spark-core. I think we need to make sure the API is compatible across all spark version we support

File: cassandra-analytics-core/src/test/spark3/org/apache/cassandra/spark/common/SidecarInstanceFactoryTest.java (hunk 1 above)
Comment (yifan-c) [reply cassandra-analytics-core/src/test/spark3/org/apache/cassandra/spark/common/SidecarInstanceFactoryTest.java:46]: Sidecar client constructor `URI` for requests. It asks for bracketed IPv6 address. 
> If a host is given then it is appended. If the host is a literal IPv6 address but is not enclosed in square brackets ('[' and ']') then the square brackets are added.

File: cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/utils/MapUtils.java (hunk 2 above)
Comment (yifan-c) [reply cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/utils/MapUtils.java]: good idea

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkSparkConf.java (hunk 4 above)
Comment (yifan-c) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkSparkConf.java]: forgot to merge after switch to `SidecarInstanceFactory`

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkSparkConf.java (hunk 3 above)
Comment (yifan-c) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/BulkSparkConf.java]: ok.

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/SidecarInstanceFactory.java (hunk 5 above)
Comment (yifan-c) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/SidecarInstanceFactory.java:47]: Not sure if I follow. What is the benefit of creating a `HostAndPort` instance here? It cannot be used to construct `SidecarInstanceImpl`

Comment (yifan-c) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/common/SidecarInstanceFactory.java:47]: Talked offline. We stick with the current implementation.

File: cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/utils/MapUtils.java [hunk 6]
```diff
@@ -239,4 +245,35 @@ public static boolean containsKey(Map<String, String> options, String key)
...
+        T deprecatedOptionValue = null;
+        if (options.containsKey(deprecated))
+        {
+            LOGGER.warn("The option: {} is deprecated. Please use {} instead. See option description for details.", deprecated, option);
+            deprecatedOptionValue = resolver.apply(deprecated);
+        }
+
+        if (options.containsKey(option) && deprecatedOptionValue != null)
+        {
+            LOGGER.info("The option: {} is defined. Favor the value over {}", option, deprecated);
```
Comment (frankgh) [inline cassandra-analytics-common/src/main/java/org/apache/cassandra/spark/utils/MapUtils.java]: Wouldn't we be skipping this section altogether when the deprecated option is not provided?

```suggestion
        if (options.containsKey(option))
        {
            if (deprecatedOptionValue != null)
            {
                LOGGER.info("The option: {} is defined. Favor the value over {}", option, deprecated);
            }
```

**Review by frankgh — CHANGES_REQUESTED** [review]

File: scripts/build-sidecar.sh [hunk 7]
```diff
@@ -24,7 +24,7 @@ else
   SCRIPT_DIR=$( dirname -- "$( readlink -f -- "$0"; )"; )
   SIDECAR_REPO="${SIDECAR_REPO:-https://github.com/apache/cassandra-sidecar.git}"
   SIDECAR_BRANCH="${SIDECAR_BRANCH:-trunk}"
-  SIDECAR_COMMIT="${SIDECAR_COMMIT:-e95786a077e1137dcaae206854986987edc6a71e}"
+  SIDECAR_COMMIT="${SIDECAR_COMMIT:-55a9efee30555d3645680c6524043a6c9bc1194b}"
```
Comment (frankgh) [inline scripts/build-sidecar.sh]: rebase maybe?

**Review by frankgh — APPROVED** [review]
+1 Looks good to me

**Pull request merged**